ports/atmel-samd/samd21_clocks.c

```diff
@@ -0,0 +1,632 @@
+                sprintf_flag(valstr, nr, 7, "WRTLOCK", SYSCTRL->XOSC32K.bit.WRTLOCK);
+                sprintf_val(valstr, nr, 8, "STARTUP", SYSCTRL->XOSC32K.bit.STARTUP);
+                if (nr > 8)
+                    return -1;
+                else
+                    return 0;
+            case 6:
+                sprintf_flag(valstr, nr, 1, "RUNSTDBY", SYSCTRL->OSC8M.bit.RUNSTDBY);
+                sprintf_flag(valstr, nr, 2, "ONDEMAND", SYSCTRL->OSC8M.bit.ONDEMAND);
+                sprintf_val(valstr, nr, 3, "PRESC", SYSCTRL->OSC8M.bit.PRESC);
+                sprintf_val(valstr, nr, 4, "CALIB", SYSCTRL->OSC8M.bit.CALIB);
+                sprintf_val(valstr, nr, 5, "FRANGE", SYSCTRL->OSC8M.bit.FRANGE);
+                if (nr > 5)
+                    return -1;
+                else
+                    return 0;
+            case 7:
+                sprintf_flag(valstr, nr, 1, "MODE", SYSCTRL->DFLLCTRL.bit.MODE);
+                sprintf_flag(valstr, nr, 2, "STABLE", SYSCTRL->DFLLCTRL.bit.STABLE);
+                sprintf_flag(valstr, nr, 3, "LLAW", SYSCTRL->DFLLCTRL.bit.LLAW);
+                sprintf_flag(valstr, nr, 4, "USBCRM", SYSCTRL->DFLLCTRL.bit.USBCRM);
+                sprintf_flag(valstr, nr, 5, "RUNSTDBY", SYSCTRL->DFLLCTRL.bit.RUNSTDBY);
+                sprintf_flag(valstr, nr, 6, "ONDEMAND", SYSCTRL->DFLLCTRL.bit.ONDEMAND);
+                sprintf_flag(valstr, nr, 7, "CCDIS", SYSCTRL->DFLLCTRL.bit.CCDIS);
+                sprintf_flag(valstr, nr, 8, "QLDIS", SYSCTRL->DFLLCTRL.bit.QLDIS);
+                sprintf_flag(valstr, nr, 9, "BPLCKC", SYSCTRL->DFLLCTRL.bit.BPLCKC);
+                sprintf_flag(valstr, nr, 10, "WAITLOCK", SYSCTRL->DFLLCTRL.bit.WAITLOCK);
+                sprintf_val(valstr, nr, 11, "FINE", SYSCTRL->DFLLVAL.bit.FINE);
+                sprintf_val(valstr, nr, 12, "COARSE", SYSCTRL->DFLLVAL.bit.COARSE);
+                sprintf_val(valstr, nr, 13, "DIFF", SYSCTRL->DFLLVAL.bit.DIFF);
+                sprintf_val(valstr, nr, 14, "MUL", SYSCTRL->DFLLMUL.bit.MUL);
+                sprintf_val(valstr, nr, 15, "FSTEP", SYSCTRL->DFLLMUL.bit.FSTEP);
+                sprintf_val(valstr, nr, 16, "CSTEP", SYSCTRL->DFLLMUL.bit.CSTEP);
+                if (nr > 16)
+                    return -1;
+                else
+                    return 0;
+            // TODO 8: FDPLL96M
+            default:
+                return -1;
+        }
+    }
+    return -1;
+}
+#else
+int clock_get_val_str(uint8_t type, uint8_t index, uint8_t nr, char *valstr) {
+    return -1;
+}
+#endif
+
+
+CLOCK(XOSC, 0, 0);
+CLOCK(GCLKIN, 0, 1);
+CLOCK(GCLKGEN1, 0, 2);
+CLOCK(OSCULP32K, 0, 3);
+CLOCK(OSC32K, 0, 4);
+CLOCK(XOSC32K, 0, 5);
+CLOCK(OSC8M, 0, 6);
+CLOCK(DFLL48M, 0, 7);
+CLOCK(FDPLL96M, 0, 8);
+
+CLOCK(DFLL48M_REF, 1, 0);
+CLOCK(DPLL, 1, 1);
+CLOCK(DPLL_32K, 1, 2);
+CLOCK(WDT, 1, 3);
+CLOCK(RTC, 1, 4);
+CLOCK(EIC, 1, 5);
+CLOCK(USB, 1, 6);
```

Comment:
Index of the bare index how about using the CMSIS macros? For USB its USB_GCLK_ID. defined in `asf4/samd21/include/instance/usb.h`